### DISCUSSION TITLE
Can unpoly exchange HTML fragments through WebSockets?

### DISCUSSION THREAD

**Discussion by longnight** [unpoly/unpoly#595] [Q&A]
Apologies for bringing up a non-issue. I want to use unpoly on my website and exchange HTML fragments with my backend through WebSockets. Does unpoly currently support doing that?

**Comment by triskweline** [comment]
Unpoly will not open a WebSocket connection for you. However, if you set up that connection yourself, you can pass HTML fragments to Unpoly to render. See [rendering strings of HTML](https://unpoly.com/render-content#local) in the docs.

**Reply by longnight** [reply]
Thank you. yes I will manage a websocket instance in "global scope" myself, then every html fragment exchange should go through it .  By using websocket do that , when there are many small HTML fragments needing updates on the page, it can reduce a lot of unnecessary HTTP request connections, and the repeated backend identity verification computations they lead to. Is my thought correct?

**Reply by triskweline** [reply]
When you already have HTML strings through your WebSocket connection, you can render them without making another HTTP request. See the doc page I linked above.

**Comment by hpvd** [comment]
just fyi:
using advanced features available in http/2 should give many/most of the efficiency and speed advantages of websocket
There was a detailed thread with many deep links to this topic in another project see https://github.com/beenotung/ts-liveview/issues/13

**Reply by hpvd** [reply]
to sum up, its mainly about bidi streaming:

- SSE to push messages down
  97% coverage at can-I-use https://caniuse.com/?search=Server%20sent%20event
- fetch api to send requests up
    96% coverage at can-I-use https://caniuse.com/mdn-api_fetch
    and there seems to be also a polyfill
    if one really!? want/need more browser support (e.g. IE10) https://github.com/github/fetch

**Reply by longnight** [reply]
Thank you for providing the links, I am amazed, I will read them carefully.

**Comment by adam12** [comment]
You don't necessarily need to use WebSockets either. It will work fine with server-side events as well.

```ruby
require "sinatra"
require "async"

get "/" do
  <<~HTML
    <!DOCTYPE html>
    <html>
    <head>
      <script type="importmap">
      {
        "imports": {
          "unpoly": "https://esm.sh/unpoly@3.7.3/unpoly.js"
        }
      }
      </script>

      <script type="module">
        import * as unpoly from "unpoly"

        const source = new EventSource('stream')
        source.onerror = () => source.close()

        source.onmessage = (e) => {
          up.render({fragment: e.data })
        }
      </script>
    </head>
    <body>
      <div id="foo">This is the original foo element</div>
    </body>
    </html>
  HTML
end

get "/stream", provides: "text/event-stream" do
  stream do |out|
    Sync do |task|
      task.async do
        sleep 10
        stream_html(out, <<~HTML)
          <div id="foo">This is the last foo</div>
        HTML
      end

      task.async do
        sleep 2
        stream_html(out, <<~HTML)
          <div id="foo">This is a new foo</div>
        HTML
      end
    end
  end
end

def stream_html(out, fragment)
  out << "data: #{fragment.gsub("\n", "")}\n\n"
end
```

**Reply by hpvd** [reply]
@adam12 many thanks for jumping in and ad this concrete SSE example !!
